### PR TITLE
Reorder semi and anti joins.

### PR DESCRIPTION
**Pull request by Tmonster**
follow up of https://github.com/duckdb/duckdb/pull/11573

To do this, we do the following,

1. When we look at reordering and what relations can join with what other relations, we treat semi joins and anti joins like regular inner joins.
2. To make sure the left and right children are not flipped, we compare the left and right bindings from the original join to the left and right bindings of the recreated plan. If the children are flipped, we flip them back.
3. When calculating the cost of a semi join we pretend the left table (i.e the table that propagates the up) is being filtered on. We take the cardinality of the left side and multiply it by 0.2. 

Since this cardinality estimation method uses multiplication, it is also symmetrical, which means we don't have to worry about different join plans for the same set of relations having different estimated cardinalities. 

Calculating the denominator of the estimated cardinality is easier now. It works like finding a maximum spanning tree.  Assuming relations are nodes and join filters are weighted edges, the process of finding the most selective filters is exactly like a maximum spanning tree problem. The weights of the edges come from the shared total domain of the columns of the filter.

Some other small improvements:
1. More checks in the cardinality estimator to make sure the denominator is not 0 (since that will cause very inaccurate cardinality estimates)
2. More checks during statistics extraction to make sure the distinct count of a column is always between 1 and the cardinality of the table.

------- 
Another heuristic was added for determining the number of distinct elements in a column as well. For integral type columns, if the maxVal - minVal is less than the distinct count measured by HLL, then DuckDB will prefer max - min as the distinct count. 

I thought this was the source of a bad join order. Turns out that wasn't the case, but I think it is still a good join heuristic.

### CODE REVIEW

**Review by lnkuiper — COMMENTED** [review]
Looks much more readable now with the refactor, thanks!

Are the changes to the `ColumnLifetimeAnalyzer` on purpose, or were you doing that on another branch?

Also, we seem to have a regression for TPC-DS q95, which has two semi-joins. This wasn't the case in your previous PR. Could you check what's going on there? Maybe the semi-join cardinality estimation heuristic is too simple?

**Review by lnkuiper — APPROVED** [review]
Thanks for the fixes, LGTM now. Our regression tests say that TPC-DS is ~2% faster now!

I took a quick peek, and these are the changed timings:
```
benchmark/tpcds/sf1/q58.benchmark
Old timing: 0.042496
New timing: 0.023607

benchmark/tpcds/sf1/q64.benchmark
Old timing: 0.261861
New timing: 0.324018

benchmark/tpcds/sf1/q70.benchmark
Old timing: 0.052156
New timing: 0.061281

benchmark/tpcds/sf1/q83.benchmark
Old timing: 0.024797
New timing: 0.011348

benchmark/tpcds/sf1/q90.benchmark
Old timing: 0.010705
New timing: 0.006925

benchmark/tpcds/sf1/q73.benchmark
Old timing: 0.030077
New timing: 0.021356
```
Looking good!

Maybe we should add TPC-DS to the plan cost at some point?

File: src/include/duckdb/optimizer/join_order/cardinality_estimator.hpp [hunk 1]
```diff
@@ -33,11 +43,29 @@ struct RelationsToTDom {
...
 };
 
+class FilterInfoWithTotalDomains {
+public:
+	FilterInfoWithTotalDomains(FilterInfo *filter_info, RelationsToTDom &relation2tdom)
+	    : filter_info(filter_info), tdom_hll(relation2tdom.tdom_hll), tdom_no_hll(relation2tdom.tdom_no_hll),
+	      has_tdom_hll(relation2tdom.has_tdom_hll) {
+	}
+
+	FilterInfo *filter_info;
```
Comment (Mytherin) [inline src/include/duckdb/optimizer/join_order/cardinality_estimator.hpp:53]: nit: Can this be a reference or an optional_ptr?

File: src/optimizer/join_order/cardinality_estimator.cpp [hunk 2]
```diff
@@ -123,44 +121,117 @@ void CardinalityEstimator::RemoveEmptyTotalDomains() {
...
+	switch (filter.filter_info->join_type) {
+	case JoinType::INNER: {
+		new_denom *= filter.has_tdom_hll ? (double)filter.tdom_hll : (double)filter.tdom_no_hll;
+		return new_denom;
 	}
+	case JoinType::SEMI:
+	case JoinType::ANTI: {
+		if (JoinRelationSet::IsSubset(*left.relations, *filter.filter_info->left_set) &&
+		    JoinRelationSet::IsSubset(*right.relations, *filter.filter_info->right_set)) {
+			new_denom = left.denom * 5;
```
Comment (Mytherin) [inline src/optimizer/join_order/cardinality_estimator.cpp:221]: Can this be a constant not a magic number?

**Review by Mytherin — COMMENTED** [review]
Looks great! Some minor nitpicks but otherwise good to go from my side

**Comment by Tmonster** [comment]
@Mytherin this ready to go now. You reviewed once already, but I had to battle CI for a while since I also needed to patch substrait

**Pull request merged**